Feature/utc 330 image hover overlay

UTCbridget opened the pull request:
This component is responsive, is available for both horizontal and vertical images and has more limited capability than cards.
 
This PR is a revision of the previous pull request regarding image hover effects. This reuses repetitive fields using the Paragraphs module. The Paragraphs module was causing issues with integration with custom blocks and layout builder, but this has been fixed with this patch: https://www.drupal.org/files/issues/2021-05-03/paragraphs-2901390-82-withtests.patch
Paragraphs now works great with custom blocks and layout builder, as well as the media module.

**Desktop Demo:**
![sample-hover-effects-desktop](https://user-images.githubusercontent.com/82905787/135518705-58ba367f-7337-4fe9-82db-c1dbf5afd34d.gif)

**Mobile Demo:**
NOTE: As in the previous PR, in the desktop version, the image itself is a link accessed through a hover (shows text) then a click (goes to the next page). On the mobile version, an actual button is needed because there is no "hover" with a finger action. Therefore, a tap on the image shows the text. The button shows to provide a means to navigating to the next page.

![sample-hover-effects-mobile2](https://user-images.githubusercontent.com/82905787/135518734-09cea224-9cc2-4449-8c23-7414eea56b7c.gif)

**Admin Interface:**
![sample-hover-effects-admin2](https://user-images.githubusercontent.com/82905787/135518795-b4f77c40-1fa3-4054-849c-4277a51a4e77.gif)

UTCbridget committed:
UTC-330: Add image hover component.

UTC-330: Add image hover component.

UTCbridget committed:
UTC-330: Add block extension for editing.

Update block--utc-image-hover-effect.html.twig with block.html.twig extension in order to be able to edit this block in layout builder.

UTCbridget committed:
UTC-330: Fix tailwind.config.js

Stray comma removed.

UTCbridget committed:
UTC-330: Fix tailwind.config.js file.

This one doesn't show, but there was code from another pull request interfering with this one. This is a restoration of original files with a comma fix.

UTCbridget committed:
UTC-330: Remove dist files from the PR.

UTC-330: Remove dist files from the PR.

UTCbridget committed:
UTC-330: Remove two more dist files from PR.

UTC-330: Remove two more dist files from PR.

UTCbridget committed:
UTC-330: Fix tailwind.config.js PR.

UTC-330: Fix tailwind.config.js PR picked up from another PR.

UTCbridget committed:
UTC-330:  Fix tailwind.config.js coupled with another PR change.

UTC-330:  My first PR incorrectly mixed my changes with another developer's changes in the tailwind.config.js. This commit includes both mine and the other developer's changes (utchighlight).

UTCbridget committed:
UTC-330: Add image hover effect using Paragraphs model.

UTCbridget committed:
UTC-330: Undo all the dist commits.

UTCbridget committed:
Merge branch 'develop' of github.com:UTCWeb/particle into feature/UTC-330-image-hover-overlay

UTCbridget committed:
UTC-330: Save Brittany's file.

UTCbridget commented:
Currently the name of the component needs work as it is too long: UTC Images with Hover Overlay
One thing I do recommend is keeping "images" plural as you can have multiple images.

Other suggestions (please feel free to offer yours!)
UTC Images Hover Effect
UTC Images Hover Overlay
UTC Images/Text Overlay
UTC Images/Text Effect

brittrich commented:
> Currently the name of the component needs work as it is too long: UTC Images with Hover Overlay One thing I do recommend is keeping "images" plural as you can have multiple images.
> 
> Other suggestions (please feel free to offer yours!) UTC Images Hover Effect UTC Images Hover Overlay UTC Images/Text Overlay UTC Images/Text Effect

I think any of those are good options. Just adding a few:

UTC Image Links?
UTC Image Link Cards?
UTC Hover Images?

UTCbridget commented:
I vote for UTC Hover Images, as that is close to what I googled when I was searching for them. @bmartinez287 ?

bmartinez287 commented:
Sounds good to me, I wonder if we should put twig files associated with it under a different folder structure.

Like:

- Component
-     Component name ( Image block, UTC hero)
-      Paragraphs
-         Paragraphs  Name  (UTC Hover Image) 
-                 Block - If there are any block twig files they go into a folder called blocks
-                 Field -  if there are any field
-                 Paragraph.twig file.

This way we keep all the fields in one folder structure.

We originally used the structure of files and folders themag and other themes had where blocks is a folder, field is a folder and so forth.
The challenge with that approach is that it places a lot of files under one folder making it harder for us to fully test what's being used or not. I removed some unused files from the blocks folder but I am sure there are a few more there that should be on ints own folder.

UTCbridget commented:
I think that's a great idea. Did you want me to do that with this component?

bmartinez287 commented:
That be great if we could do it here. It be the start of it. Allowing us to associate them a bit better.

brittrich commented on apps/drupal-default/particle_theme/templates/block/block--utc-image-hover-effect.html.twig:
```diff
@@ -0,0 +1,106 @@
+		hover_text: content.field_hover_image_three_body|field_value,
+		hover_link: content.field_hover_image_three_link.0['#url'],
+    } %}
+
```
@UTCbridget apologies if I'm spacing on it (possible ;))...but is this file an older version that will be removed? Looks like the updated paragraphs files are handling this now with less fields?

UTCbridget replied:
Yes. I'm using paragraphs. I got hung up with some TDX tickets. I'll get back to this hopefully by week's end.

UTCbridget replied:
Paragraphs is working great now that we applied the patch.

UTCbridget committed:
UTC-341: Update image hover by moving all related files into one folder.

UTCbridget committed:
UTC-341: Remove old files now we're consolidated.

UTCbridget committed:
UTC-341: Remove more old files.

UTCbridget commented:
@bmartinez287, I'm still working on this. This is not ready at this moment. Please wait until I let you know before reviewing and merging.

UTCbridget committed:
UTC-330: Image hover update.

UTCbridget committed:
UTC-330: remove old files.

UTCbridget committed:
UTC-330: Remove tailwind.config.js as it is not changed for this PR.

UTCbridget committed:
UTC-330: Remove tailwind.config.js from this PR again.

UTCbridget committed:
UTC-330: Update image hovers with reponsive tweaks.

UTCbridget committed:
UTC-330: Added comments to js file.

UTCbridget commented:
This is finally ready for review...

This latest version of Image Hovers uses paragraphs in order to reuse fields. 

CMS Notes:
---If the user does not want the image to be a link, the cursor is set to default.
---Max Length has been applied as the allowable space is constricted and cannot grow.
---The CMS is well-commented for clear instructions for the user.
---Titles are required. Body copy is not.

CSS is applied to address the following conditions:
---1-5 images in a row
---horizontal and vertical orientations
---resizing hz and vt images when user combines them in a row
---addition of button links for mobile devices.

**Desktop examples:**

### Horizontal
![hover-image-hz-desktop](https://user-images.githubusercontent.com/82905787/139724548-29f15eaa-eb9f-4195-9980-6b300cc0cba2.gif)

### Vertical
![hover-image-vt-desktop](https://user-images.githubusercontent.com/82905787/139724558-886b3187-5361-48f7-b8d7-6a3e35228d67.gif)

### CMS:
_Note: Up to five images can be added. Instructions are detailed. Users are warned that if image orientation do not match the required row selection, images will be resized and cropped._

![hover-image-admin](https://user-images.githubusercontent.com/82905787/139724665-63bb229a-bad3-4d68-bca9-4136fa165f35.gif)

### Example of mixed orientations added by the user:
_Note: The center image in each row does not match the row's orientation, therefore, the image's aspect ratio has to be changed, resulting in the image being cropped._

![hover-image-mix-desktop](https://user-images.githubusercontent.com/82905787/139725420-deb6a23a-d251-45e7-b7df-0182db6722c3.gif)

**iPad examples**

### Horizontal:
![hover-image-hz-ipad](https://user-images.githubusercontent.com/82905787/139725664-8244ae9b-8e92-46c8-94a8-4c3305a08643.gif)

### Vertical:
![hover-image-vt-ipad](https://user-images.githubusercontent.com/82905787/139725724-42da51f0-5df3-40a6-be16-bcc8a08db23b.gif)

**iPhone examples:**

### Horizontal:
![hover-image-hz-mobile](https://user-images.githubusercontent.com/82905787/139725829-d3f3e08b-3d22-4c87-a906-0afedcf56740.gif)

### Vertical:
![hover-image-vt-mobile](https://user-images.githubusercontent.com/82905787/139725850-05cfd553-b36e-4f6a-996d-8f9fa4e40167.gif)

bmartinez287 commented on source/default/_patterns/00-protons/legacy/js/utc-hover-images.js:
```diff
@@ -0,0 +1,47 @@
+(function($, Drupal, drupalSettings) {
+    "use strict";
+   //Adds slick arrows 
+    Drupal.behaviors.slickcustomarrows = {
```
slickarrows are for when it goes mobile?

UTCbridget replied:
Oh sorry, that's a cut an paste error. Let me fix it.

UTCbridget committed:
UTC-330: Correct hover images js file.

UTCbridget committed:
UTC-330: Delete block--block--image-with-hover-overlay.html.twig duplicate.

bmartinez287 commented on apps/drupal-default/particle_theme/templates/paragraphs/image-hover/block--image-with-hover-overlay.html.twig:
```diff
@@ -0,0 +1,36 @@
+#}
+
+{% block content %}
+	{% set rendered_content = content|render %}
```
It seems like this variable is not used. Do we need it?

bmartinez287 commented on apps/drupal-default/particle_theme/templates/paragraphs/image-hover/block--image-with-hover-overlay.html.twig:
```diff
@@ -0,0 +1,36 @@
+
+{% block content %}
+	{% set rendered_content = content|render %}
+    {% set hover_image_orientation = content.field_image_orientation|field_value|render %}
```
What about this one?

bmartinez287 commented on apps/drupal-default/particle_theme/templates/paragraphs/image-hover/paragraph--hover-images.html.twig:
```diff
@@ -0,0 +1,75 @@
+              <div class="utc-border-wrapper">
+                <h2 class="m-0 lg:mb-2 font-bold capitalize text-utc-new-blue-500 leading-tight">{{ content.field_hover_title.0 }}</h2>
+                <p class="m-0 mb-2 text-utc-new-blue-500 leading-tight">{{ content.field_hover_text.0 }}</p>
+                <p class="hover-image-space">&nbsp;</p>
```
why do we have an empty p tag?

UTCbridget replied:
Because the anchor on the mobile turns into a button link, instead of the whole text area being the link, and I need to add space between the text and the button. The margin-bottom in the paragraph above won't apply here (and is removed in my new files), because of the transitions. I can remove it and use another solution.

UTCbridget committed:
UTC-341: Update files with cleaner css and better application to internal pages.

UTCbridget commented:
This commit provides cleaner css as well as css support for internal page applications, which was lacking before.

bmartinez287 commented on source/default/_patterns/00-protons/legacy/css/components/UTC-custom-blocks/_utc_hover_images.css:
```diff
@@ -0,0 +1,341 @@
+
+
+/****Internal pages treatment***/
+.themag-layout--twocol-section--3-9 figure.utc-hover-image-effect figcaption {
```
the challenge with this it assumes the 3-9 layout but they could flip it to 3-9 or in some cases is a 3 column layout.

bmartinez287 commented on apps/drupal-default/particle_theme/templates/paragraphs/image-hover/block--utc-hover-images.html.twig:
```diff
@@ -0,0 +1,36 @@
+{% extends directory ~ "/templates/block/block.html.twig" %}
```
I never seen directory or ~. 
I think you could do just `{% extends "block.html.twig" %}`
That may be causing some issue although its working somewhat.

UTCbridget replied:
Because we moved all the twig files for this component into one folder, the block twig is no longer in the block folder where that extended file resides. This is the code that tells Drupal where to find that extended file. I have to have this extension in order to edit it in Layout Builder.

UTCbridget committed:
UTC-330: Fix layout builder's drag and drop issue.

UTCbridget commented:
I have fixed the drap and drog ;) issue by remove auto attributes duplicated to a child div in the block override twig. The file affected is this: apps/drupal-default/particle_theme/templates/paragraphs/image-hover/block--utc-hover-images.html.twig

UTCbridget committed:
UTC-330: This removed the media twig override.

UTCbridget commented:
I removed the custom media twig file that was applied to this component. The fix is in the database through changing the Hover Image field from "Rendered Entity" in the Paragraphs > Manage Display to "Field Formatter with Inline Settings."

UTCbridget committed:
UTC-330: Remove path directory from block extension call.

UTCbridget committed:
UTC-341: Remove 4 files, consolidate code in block twig, and update css for image hover.

UTCbridget committed:
UTC-341: Remove the js file from index.js.

UTCbridget committed:
UTC-341: Move the hover image block twig to the block folder.

Since we no longer have to group hover image files together, we move this file to where the other block twig files are.

UTCbridget commented:
This commit does the following:

A. Removes three twig files: 
1. paragraph--hover-images.html.twig
2. field--field-image-orientation.html.twig
3. field--field-hover-image-block.html.twig

B. Edit the block--utc-hover-images.html.twig
1. Consolidates the A. files' code into this block twig.
2. Adds vertical and horizontal configuration from the image style implementations.
3. Moves this file to the block folder since there are no other paragraph twigs to keep together anymore.

C. Removes the need for js, removing the utc-hover-images.js file and it's call in index.js.

D. Updates the _utc_hover_images.css with the removal of aspect ratio implementation.

brittrich commented:
@UTCbridget is this ready to review with the config? I was going to do an actual pull to test...but wanted to be sure it was updated in both spots first. Looks great, just wanted to try it IRL. :)

UTCbridget commented:
@brittrich, almost, tweaking the css. Bernardo and I went over it a couple of days ago and found a couple of glitches, which I fixed. Should have it done today (11/18/21). I will ping you in Teams. I managed to put all the code in one twig file instead of the 4 twig files I originally had, and got rid of the js file altogether. So now I'm down to three files, the block twig, the css, and the index.js for the css.

UTCbridget committed:
UTC-341: Fix link issue and fine-tune the css for responsiveness.

bmartinez287 commented:
@UTCbridget are we good to go?  I was going to merge into particle so we can have an easier time testing it and make any last minute changes as a separate PR. Let me know.

UTCbridget commented:
Yes, @bmartinez287, this PR is good to go. I have to update the Drupal core per your comments and [re-PR the utccloud config files](https://github.com/UTCWeb/utccloud/pull/1931) though.
